include/boost/gil/algorithm.hpp

```diff
@@ -1174,6 +1180,423 @@ F transform_pixel_positions(const View1& src1,const View2& src2,const View3& dst
     }
     return fun;
 }
+
+
+// Code below this line is moved here from <boost/gil/extension/numeric/algorithm.hpp>
+
+/// \brief Reference proxy associated with a type that has a \p "reference" member type alias.
+///
+/// The reference proxy is the reference type, but with stripped-out C++ reference.
+/// Models PixelConcept.
+template <typename T>
```

Comment:
The whole core `algorithm.hpp` is becoming huge!

What about that we start splitting it:
1. `gil/algorithm.hpp` to `gil/algorithm/core.hpp`
2. `gil/extension/numeric/algorithm.hpp` to `gil/algorithm/numeric.hpp`

Next, we could start splitting `gil/algorithm/core.hpp` into better structured/categorised headers

Reply:
Yes, that makes sense and we should definitely do that. but on the other hand it also gets confusing to me(as a user) from header names it is not clear what will be in what. and then also we have image processing ALGORITHMS which will not be in the algorithm directory. so I am not sure how to proceed with this. or maybe another major reflector and rearranging of algorithms?

Reply:
Rearranging of algorithms is something on its way too, I think.
That is why it would be nice to prepare a clear structure for it.

My idea was to use directory and file names as from general to specific.
There obviously are at least two approaches/directions:

1.  `gil/algorithm/<what kind of algorithms>.hpp` or `gil/algorithm/<name of algorithm>.hpp`  with `gil/algorithm/core.hpp` are general purpose/basic algorithms.
2.  `gil/image_processing/algorithm.hpp`, `gil/rasterisation/algorithm.hpp`, `gil/numeric/algorithm.hpp`,  etc.

To be honest, I'm more in favour of the option 1. as it would lead us to filesystem with clear separation of data and algorithms.

> we have image processing ALGORITHMS which will not be in the algorithm directory

Good point. IMO, IP algorithms could be moved too.
TBH,  `image_processing` as folder name is an unfortunate one (blame on me, I think) as the whole library is about image processing, isn't it :-)

Reply:
agreed. though I would like to keep that in a separate PR/issue and get done with this PR. 

ideal steps to move forward IMO:
1. get this PR done
2. we refactor the codebase (doing refactor while large PRs are waiting is painful so before or after large PRs)
3. I start rearranging the algorithms

Reply:
Sure, I'm just brainstorming. This is your PR, so you're in judge of its scope.

Reply:
reviews appreciated :)